contrib/epee/include/net/net_ssl.h

```diff
@@ -129,6 +129,24 @@ namespace net_utils
         \return True if the SSL handshake completes with peer verification
           settings. */
     bool handshake(boost::asio::ssl::stream<boost::asio::ip::tcp::socket> &socket, boost::asio::ssl::stream_base::handshake_type type, const std::string& host = {}) const;
+
+    //! See `handshake` description
+    template<typename HandshakeHandler>
+    void async_handshake(
+      boost::asio::ssl::stream<boost::asio::ip::tcp::socket> &socket,
+      boost::asio::ssl::stream_base::handshake_type type,
+      const std::string& host,
+      HandshakeHandler &&handler) const
+    {
+      configure_verification(socket, type, host);
```

Comment:
The `host` is used in the `verify_callback`, which means the reference must remain valid until `handler` is invoked. This is the case in current uses, but is easy to miss. At a minimum this needs to be documented.

More realistically, the values should be taken by value in that lambda OR the business logic loop above should be moved into the `handshake` function within the cpp so that all uses still block. If the caller starts a timer, the effect is the same, and it simulatenously fixes the same issue in the epee TCP server when using one I/O thread.

Reply:
> The host is used in the verify_callback, which means the reference must remain valid until handler is invoked. This is the case in current uses, but is easy to miss. At a minimum this needs to be documented.

Won't fix, Is not an issue. The same is applicable to other code parts using async functions as well, think of async reads and writes and the buffers you passing in.
Does this bother you as well?
If so, you better submit a patch with changes you want to all the async functions at once.

>  the business logic loop above should be moved into the handshake function within the cpp so that all uses still block. If the caller starts a timer, the effect is the same, and it simulatenously fixes the same issue in the epee TCP server when using one I/O thread.

TCP server should use `async_handshake` instead of `handshake`, this will fix TCP server issue.
It is out of scope for the current PR. 
As always, patches from anyone eager to fix TCP server part are welcome.

Reply:
> Won't fix, Is not an issue. The same is applicable to other code parts using async functions as well, think of async reads and writes and the buffers you passing in.
> Does this bother you as well?
> If so, you better submit a patch with changes you want to all the async functions at once.

Yes, actually. People frequently get the behavior wrong. Its why other languages have moved to "fibers" (go threads), etc., because its easy to botch this kind of thing.

There are advantages to this implementation, but no uses of this function are actually async currently. The epee server is still explicitly blocking and the client code is also blocking until the timer fires. So until the epee server is refactored to be fully async (which is doable), there isn't even a need to for this async functionality (which is easy to re-introduce).

However, asio does document the behavior. As I originally stated, please at least document the restrictions in the header file, because its not obvious by the interface design.

Reply:
Alright, don't like half measures either (i.e. fixing only client, but no server).

Will patch `handshake` routine so it will accept `timeout` parameter and will have its own deadline timer inside. Don't want to rely on the `blocked_mode_client` timer because `ssl_options_t` knows nothing about it.